src/libsyntax/parse/parser.rs

```diff
@@ -6519,11 +6553,15 @@ impl<'a> Parser<'a> {
                                visibility: Visibility,
                                attrs: Vec<Attribute>)
                                -> PResult<'a, P<Item>> {
-        let orig_name = self.parse_ident()?;
+        // Accept `extern crate name-like-this` for better diagnostics
+        let ident = self.parse_crate_name_with_dashes(
+            "crate name using dashes are not valid in `extern crate` statements",
```

Comment:
typo: name -> names

Nit 1: could you keep `orig_name` instead of `ident`, this is a point of confusion because the meaning of optional identifier in `ItemKind::ExternCrate` is a bit counter-intuitive.

Nit 2: The messages can be moved into `parse_crate_name_with_dashes`.